File: src/main/java/me/testaccount666/serversystem/moderation/BanModeration.java
package me.testaccount666.serversystem.moderation;

import lombok.experimental.SuperBuilder;

import java.util.UUID;

@SuperBuilder
public class BanModeration extends AbstractModeration {
    public BanModeration(long issueTime, long expireTime, String reason, UUID senderUuid, UUID targetUuid) {
        super(issueTime, expireTime, reason, senderUuid, targetUuid);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/moderation/BanModeration.kt
package me.testaccount666.serversystem.moderation

class BanModeration : AbstractModeration {

    private constructor(b: BanModerationBuilder<*>) : super(b)

    abstract class BanModerationBuilder<B : BanModerationBuilder<B>> : AbstractModerationBuilder<BanModeration, B>() {
        abstract override fun self(): B

        abstract override fun build(): BanModeration

        override fun toString() = "BanModeration.BanModerationBuilder(super=${super.toString()})"
    }

    private class BanModerationBuilderImpl : BanModerationBuilder<BanModerationBuilderImpl>() {
        override fun self() = this

        override fun build() = BanModeration(this)
    }

    companion object {
        @JvmStatic
        fun builder(): BanModerationBuilder<*> = BanModerationBuilderImpl()
    }
}
